hr_expense_from_purchase_request/models/hr_expense.py

```diff
@@ -0,0 +1,168 @@
+# Copyright 2021 Ecosoft Co., Ltd. (http://ecosoft.co.th)
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl).
+
+from odoo import api, fields, models
+
+
+class HrExpenseSheet(models.Model):
+    _inherit = "hr.expense.sheet"
+
+    employee_user_id = fields.Many2one(
+        related="employee_id.user_id",
+        readonly=True,
+    )
+    purchase_request_id = fields.Many2one(
+        comodel_name="purchase.request",
+        string="Purchase Request",
+        ondelete="restrict",
+        copy=False,
+        index=True,
+        help="Select purchase request of this employee, to create expense lines",
+    )
+    pr_for = fields.Selection(
+        selection=[("expense", "Expense")],
+        string="Use PR for",
+        default="expense",
+        required=True,
+    )
+    pr_line_ids = fields.One2many(
+        comodel_name="hr.expense.sheet.prline",
+        inverse_name="sheet_id",
+        copy=False,
+    )
+
+    @api.onchange("employee_id")
+    def _onchange_purchase_request_employee(self):
+        self.purchase_request_id = False
+
+    @api.onchange("purchase_request_id")
+    def _onchange_purchase_request_id(self):
+        SheetPRLine = self.env["hr.expense.sheet.prline"]
+        self.pr_line_ids = False
+        for line in self.purchase_request_id.line_ids:
+            sheet_prline = self._prepare_sheet_prline(line)
+            self.pr_line_ids += SheetPRLine.new(sheet_prline)
+
+    def _prepare_sheet_prline(self, line):
+        """ Prepare data, to create hr.expense. All must be hr.expense's fields """
+        unit_amount = (
+            line.estimated_cost / line.product_qty if line.product_qty > 0 else 0
+        )
+        return {
+            "name": line.name,
+            "product_id": line.product_id.id,
+            "product_uom_id": line.product_uom_id.id,
+            "unit_amount": unit_amount,
+            "quantity": line.product_qty,
+            "company_id": line.company_id.id,
+            "currency_id": line.currency_id.id,
+            "analytic_account_id": line.analytic_account_id.id,
+            "analytic_tag_ids": line.analytic_tag_ids.ids,
+            "description": line.description,
+            "reference": line.specifications,
+        }
+
+    @api.model
+    def create(self, vals):
+        sheet = super().create(vals)
+        if "purchase_request_id" in vals:
+            sheet.mapped("expense_line_ids").unlink()
+        sheet._do_process_from_purchase_request()
+        sheet.pr_line_ids.unlink()  # clean after use
+        return sheet
+
+    def write(self, vals):
+        res = super().write(vals)
+        if "purchase_request_id" in vals:
+            self.mapped("expense_line_ids").unlink()
+        self._do_process_from_purchase_request()
+        self.mapped("pr_line_ids").unlink()  # clean after use
```

Comment:
Make sure create expense when Purchase Request != False

Please change to:
```
if vals.get("purchase_request_id"):
     self.mapped("expense_line_ids").unlink()
     self._do_process_from_purchase_request()
     self.mapped("pr_line_ids").unlink()
```

Reply:
Are you sure? It was my intention to always do the job.

Reply:
Yes, I'm sure

This was example the error
1. No line when save expense report
![Peek 2021-05-26 10-22](https://user-images.githubusercontent.com/24691983/119598035-6054fe80-be0c-11eb-9088-e9c037ae80c1.gif)
2. Select PR and use pr for = Advance
![Peek 2021-05-26 10-29](https://user-images.githubusercontent.com/24691983/119598648-85963c80-be0d-11eb-8366-9ad36c733864.gif)

Reply:
I still don't face your issue. I can still use EX as normal (no line deleted). I will keep your finding in mind, if I saw it I will fix it.

Reply:
Fixed, thanks for reporting.